pkg/asset/ignition/bootstrap/gcp/storage.go

```diff
@@ -0,0 +1,152 @@
+func NewStorageClient(ctx context.Context) (*storage.Client, error) {
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create client: %w", err)
+	}
+
+	return client, nil
+}
+
+// CreateBucketHandle will create the bucket handle that can be used as a reference for other storage resources.
+func CreateBucketHandle(ctx context.Context, bucketName string) (*storage.BucketHandle, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*1)
+	defer cancel()
+
+	client, err := NewStorageClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create storage client: %w", err)
+	}
+	return client.Bucket(bucketName), nil
+}
+
+// CreateStorage creates the gcp bucket/storage. The storage bucket does Not include the bucket object. The
+// bucket object is created as a separate process/function, so that the two are not tied together, and
+// the data stored inside the object can be set at a later time.
+func CreateStorage(ctx context.Context, ic *installconfig.InstallConfig, bucketHandle *storage.BucketHandle, clusterID string) error {
+	labels := map[string]string{}
+	labels[fmt.Sprintf("kubernetes-io-cluster-%s", clusterID)] = "owned"
+	for _, label := range ic.Config.GCP.UserLabels {
+		labels[label.Key] = label.Value
+	}
+
+	bucketAttrs := storage.BucketAttrs{
+		UniformBucketLevelAccess: storage.UniformBucketLevelAccess{
+			Enabled: true,
+		},
+		Location: ic.Config.GCP.Region,
+		Labels:   labels,
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*1)
+	defer cancel()
+
+	if err := bucketHandle.Create(ctx, ic.Config.GCP.ProjectID, &bucketAttrs); err != nil {
+		return fmt.Errorf("failed to create bucket: %w", err)
+	}
+	return nil
+}
+
+// CreateSignedURL creates a signed url and correlates the signed url with a storage bucket.
+func CreateSignedURL(handle *storage.BucketHandle, objectName string) (string, error) {
+	opts := storage.SignedURLOptions{
+		Method:  "GET",
+		Expires: time.Now().Add(time.Minute * 60),
+	}
+
+	ctx := context.Background()
+	session, err := gcpic.GetSession(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to get gcp session: %w", err)
+	}
+
+	// TODO: make sure all cases are handled including the cases required by https://github.com/openshift/installer/pull/7697
+	if session.Credentials.JSON != nil {
+		var credsMap map[string]interface{}
+		if err := json.Unmarshal(session.Credentials.JSON, &credsMap); err != nil {
+			return "", err
+		}
+
+		clientEmail, ok := credsMap["client_email"]
+		if !ok {
+			return "", fmt.Errorf("failed to process client_email from gcp credentials")
+		}
+		opts.GoogleAccessID, ok = clientEmail.(string)
+		if !ok {
+			return "", fmt.Errorf("failed to convert client_email to a string")
+		}
+
+		privateKey, ok := credsMap["private_key"]
+		if !ok {
+			return "", fmt.Errorf("failed to process private_key from gcp credentials")
+		}
+		opts.PrivateKey = []byte(privateKey.(string))
+	}
```

Comment:
I think the ` storage.SigningSchemeV4` from:
https://cloud.google.com/storage/docs/access-control/signing-urls-with-helpers#client-libraries
will handle most of this for us, but on the other hand it requires a new permission `iam.serviceAccounts.signBlob` for the case when a private key is missing. 

So we need to consider this carefully for how we will backport it. By requiring a new permission, we could break existing installs. On the other hand, I think there is a decent argument to be made that it may be a necessity: in this edge case where a private key is not included in the credentials.